lib/code_corps/donation_goals_manager.ex

```diff
@@ -0,0 +1,114 @@
+defmodule CodeCorps.DonationGoalsManager do
+  import Ecto.Query
+
+  alias CodeCorps.{DonationGoal, Project, Repo}
+  alias Ecto.Multi
+
+  def create(attributes) do
+    changeset = %DonationGoal{} |> DonationGoal.create_changeset(attributes)
+
+    multi = Multi.new
+    |> Multi.insert(:donation_goal, changeset)
+    |> Multi.run(:update_current_goal, &transactional_update_current_goal/1)
+
+    case Repo.transaction(multi) do
+      {:ok, %{donation_goal: donation_goal, update_current_goal: _}} ->
+        {:ok, Repo.get(DonationGoal, donation_goal.id)}
+      {:error, :donation_goal, %Ecto.Changeset{} = changeset, %{}} ->
+        {:error, changeset}
+      {:error, _failed_operation, _failed_value, _changes_so_far} ->
+        {:error, :unhandled}
+    end
+  end
+
+  def update(donation_goal, attributes) do
+    changeset = donation_goal |> DonationGoal.create_changeset(attributes)
+
+    multi = Multi.new
+    |> Multi.update(:donation_goal, changeset)
+    |> Multi.run(:update_current_goal, &transactional_update_current_goal/1)
+
+    case Repo.transaction(multi) do
+      {:ok, %{donation_goal: donation_goal, update_current_goal: _}} ->
+        {:ok, Repo.get(DonationGoal, donation_goal.id)}
+      {:error, :donation_goal, %Ecto.Changeset{} = changeset, %{}} ->
+        {:error, changeset}
+      {:error, _failed_operation, _failed_value, _changes_so_far} ->
+        {:error, :unhandled}
+    end
+  end
+
+  def update_current_goal(%Project{} = project) do
+    unset_all_goals_from_current(project)
+    set_new_current_goal(project)
+  end
+
+  defp transactional_update_current_goal(%{donation_goal: %DonationGoal{project_id: project_id}}) do
+    Project
+    |> Repo.get(project_id)
+    |> update_current_goal
+  end
+
+  defp unset_all_goals_from_current(%Project{id: project_id}) do
+    from(d in DonationGoal, where: d.project_id == ^project_id)
+    |> Repo.update_all(set: [current: false])
+
+    {:ok, :goals_unset}
+  end
+
+  defp set_new_current_goal(%Project{} = project) do
+    project |> find_current_goal |> set_to_current
+  end
+
+  defp find_current_goal(%Project{} = project) do
+    amount_donated = get_amount_donated(project)
+
+    find_smallest_not_yet_reached(project, amount_donated)
+    |> fallback_to_largest_reached(project, amount_donated)
+  end
+
+  defp get_amount_donated(%Project{id: project_id}) do
+    # TODO: This should be simplified by having
+    # subscriptions relate to projects instead of plans
+    # and by caching the total amount on the project itself
+
+    CodeCorps.StripeConnectPlan
+    |> Repo.get_by(project_id: project_id)
```

Comment:
Same thing here. I find it difficult to read when you are piping the model into here.